src/metalog.h

```diff
@@ -183,6 +183,7 @@ typedef enum LogLineType_ {
 #define DEFAULT_UPD_PORT "514"              /* UDP port of the remote syslog server */
 #define DEFAULT_DNS_LOOKUP_INTERVERVAL 120  /* seconds, after that a DNS lookup should be repeated */
 
+/* "DEL" character or all characters < "SPACE" (< 0x20) */
```

Comment:
It might read better as `/* "DEL" (0x7f) character or all characters < "SPACE" (0x20) */`. And then change the 32U in the code to 0x20, or use 32U in the comment.

Reply:
Will do. 

Can you explain the subsitution of all unicode bytes without bits 6 and 7 set?

In NEWS there is a entry for version 1:
 - add --with-unicode option to control unicode mangling #2669923
Can you find this bug number? It obviously is not a Gentoo bug ticket.

Reply:
I have no idea what that bug number is.

The 0x60 check of bits 5 and 6 in `# define ISCTRLCODE(X) ((X) == 0x7f || !(((unsigned char) (X)) & 0x60))` matches the C0 set in ASCII (0x00-0x1F) and the C1 set in extended ASCII (0x80-0x9F). The 
ACCEPT_UNICODE_CONTROL_CHARS macro is technically incorrect because unicode includes both of those control character sets. However, metalog would need to parse each character to determine if a byte was a C1 control code or a utf8 extended byte (0x80-0xBF).

Reply:
Eventually I'm a little bit stupid? (pun intended :-)

Why bits 5 and 6 and not 6 and 7? In UTF-8 all bytes have bit 7 set to 1 except the first 127 ones, that are encoded identically like in ASCII.
All other bytes are start bytes or following bytes (the second, third or fourth byte of the character). If one would want to encode characters in the forbidden C0 or C1 set:
- two bytes would be necessary
- the start byte would be "1100 0000" (0xC0) or "1100 0001" (0xC1)
- the following (second) byte would have to start with "10xx xxxx".

The code in metalog should filter the start bytes 0xC0 and 0xC1 and also the following byte, shouldn't it?

Reply:
C0 and C1 are names for the two 32 character control code sets, not 0xC0 and 0xC1 values.

The 0x20 in the first macro strips C0 characters from utf8 sequences.

The 0x60 is for ASCII and ISO8859, not unicode. 000x_xxxx and 100x_xxxx are the C0 and C1 control code sets. Any non-control character would have bit 5 and/or 6 set.

The 0x80-0x9F control codes in the C1 set would produce the utf8 bytes 1100_0010 + 100x_xxxx which is 0xC2 + 0x80-0x9F. The 0x60 macro would see that second byte as a control code, which is fine because it is one, but it would not replace the 0xC2 byte, resulting in an invalid utf8 sequence. It would also do the same for 3+ byte sequences whose extended bytes fell in the 0x80-0x9F range. That is the "mangling" mentioned in the NEWS entry.

Metalog could replace the 0xC2 byte if the following byte was a 0x80-0x9F, but then ISO8859 logs would have valid 0xC2 `Â` characters removed if followed by a C1 control code.

> The code in metalog should filter the start bytes 0xC0 and 0xC1 and also the following byte, shouldn't it?

No. Is there code doing that?

Reply:
Thank you for the explanation!

My mistake was that I mixed up the ifdef and else macros. The first macro is for UTF8, not the second one. It is unfortunate, because in UTF-8 there is also a set of characters called "C0" and "C1", but they represent the ASCII characters from 0 to 127 and not the control groups you are talking about. These UTF-8 characters are forbidden! Metalog is not filtering these forbidden character sets. Even if you do not read German, have a look at the first coloured table: https://de.wikipedia.org/wiki/UTF-8. The red entries are not allowed. 

What I couldn't grasp is, how 0x60 could be related to these! It isn't. Sorry, obviously I didn't read your explanation thouroghly enough.

My guess is, we should leave it as it is. We could filter the forbidden UTF-8 characters, but users still could send garbage e.g. binary files.

Reply:
After getting some sleep, I now understand what you were describing. You can encode '\0' as 0x00, 0xC0+0x80, 0xE0+0x80+0x80, and so on. The utf8 encoder I wrote wouldn't produce such sequences but it would decode them just fine because detection would only slow it down. So yes, 0xC0 and 0xC1 bytes should not exist in properly encoded utf8. They can however exist in iso8859 for the 'À' and 'Á' characters.

Should the first macro mention that it does not strip C1 control characters, as that could cause issues with mixed utf8 and iso8859 logs?